### DISCUSSION TITLE
字体注册错误

### DISCUSSION THREAD

**Discussion by qingpingwang** [Tencent/libpag#2166] [General] [answered]
版本：release4.3
需求：还原ae效果；
步骤：
1.load pag；
2.遍历pag文本图层，注册字体；
3.渲染pag；
现象：
注册的字体并未使用，必须调用pagtextlayer->textDocumentForWrite()；才能通知重绘。
期望：
像前几个版本一样，支持font动态加载。不然我每次注册字体，都得调用类似layer->setText(layer->text())

**Comment by kevingpqi123** [comment, marked as answer]
没有看懂你要表述的是啥，字体注册的正确使用姿势：
1、如果需要显示特定字体，都需要通过 PAGFont 的接口注册字体然后传入到 PAGTextLayer 中，或者你明确知道系统字体的 fontFamily 和 fontStyle，通过直接赋值也可以
2、如果没有做步骤1 或步骤 1 没有匹配到相关字符，通过 PAGFont SetFallbackFontPaths 注册了回退字体，如果查找的相关字符，会使用回退字体中的一种，具体是哪个不确定；
3、如果步骤1和2 都没有做或步骤 1、2 没有匹配到相关字符，pag 内部注册了部分系统字体作为回退字体，会使用回退字体中一个，匹配到那个不确定

**Reply by qingpingwang** [reply]
我的意思，假设设计的pag用的xxx字体
我想还原这个效果，我现在的步骤必须是：
1.注册xxx字体；
2.加载pag；
3.draw；
而我实际上想做的是：
1.加载pag；
2.遍历pag，注册每个图层需要的字体；
3.draw;
这种方式就不行了，因为我实际上没有更改任何文本图层的内容，无法调用到textLayer->notifyModified(true);必须手动调用一边textDocumentForWrite才行；

**Reply by qingpingwang** [reply]
每次注册完字体，我遍历pagfile的所有文本图层，调用textLy->setText(textLy->text());已规避此问题；

**Reply by kevingpqi123** [reply]
你所谓的注册字体是怎么实现的哈，这个问题你没有说清楚：
比如 C++ 的接口：
for (int i = 0; i < pagFile->numTexts(); i++) {
    auto textDocumentHandle = pagFile->getTextData(i);
    textDocumentHandle->text = "hah哈 哈哈哈哈👌";
    // Use special font
    auto pagFont = pag::PAGFont::RegisterFont("../../resources/font/NotoSansSC-Regular.otf", 0);
    textDocumentHandle->fontFamily = pagFont.fontFamily;
    textDocumentHandle->fontStyle = pagFont.fontStyle;
    pagFile->replaceText(i, textDocumentHandle);
  }
  
或者将 pagFont 的 fontFamily 和 fontStyle 赋值给 PAGTextLayer， 这才是注册特定字体的正确实现，然后执行 flush 或 play 方法就可以看到正确结果
